File: zinc-flow-csharp/ZincFlow/Fabric/Registry.cs
using ZincFlow.Core;

namespace ZincFlow.Fabric;

/// <summary>
/// ProcessorFactory: creates a configured IProcessor from scoped context and config.
/// </summary>
public delegate IProcessor ProcessorFactory(ScopedContext ctx, Dictionary<string, string> config);

/// <summary>
/// Kind of a processor parameter — drives UI form rendering and round-trip
/// string encoding. Values are serialized to JSON as strings (see ApiHandler).
/// </summary>
public enum ParamKind
{
    String,         // single-line text
    Multiline,      // textarea
    Integer,
    Number,         // double
    Boolean,        // "true" / "false"
    Enum,           // choice list (Choices populated)
    Expression,     // EL expression — UI can syntax-highlight
    KeyValueList,   // repeater of (key, value) rows; see ValueKind + delims
    StringList,     // repeater of single-string rows (no key)
    Secret,         // env-var reference like ${MY_SECRET}
}

/// <summary>
/// Typed description of a single processor config parameter. Drives the UI
/// form for visual programming. <see cref="Default"/> distinguishes
/// null ("no default — leave blank") from "" ("default is the empty string").
/// </summary>
public sealed class ParamInfo
{
    public string Name { get; init; } = "";
    public string Label { get; init; } = "";
    public string Description { get; init; } = "";
    public ParamKind Kind { get; init; } = ParamKind.String;
    public bool Required { get; init; }
    public string? Default { get; init; }
    public string? Placeholder { get; init; }
    public IReadOnlyList<string>? Choices { get; init; }
    public ParamKind? ValueKind { get; init; }
    public string EntryDelim { get; init; } = ";";
    public string PairDelim { get; init; } = "=";
}

/// <summary>
/// Metadata about a registered processor type.
/// </summary>
public sealed class ProcessorInfo
{
    public string Name { get; }
    public string Description { get; }
    public string Category { get; }
    public IReadOnlyList<ParamInfo> Parameters { get; }
    public IReadOnlyList<string> ConfigKeys { get; }
    /// Opt-in wizard id — when set, the UI renders the named wizard
    /// component instead of the generic per-kind form. Null means use
    /// the generic form (default).
    public string? WizardComponent { get; init; }

    public ProcessorInfo(string name, string description, string category, IReadOnlyList<ParamInfo> parameters)
    {
        Name = name;
        Description = description;
        Category = category;
        Parameters = parameters;
        ConfigKeys = parameters.Select(p => p.Name).ToList();
    }

    // Legacy constructor — kept permanently so third-party processors using the
    // old (name, description, List<string>) shape still compile. Wraps each key
    // as a String-kind param with Category="Other".
    public ProcessorInfo(string name, string description, List<string> configKeys)
        : this(name, description, "Other",
               configKeys.Select(k => new ParamInfo { Name = k, Label = k, Kind = ParamKind.String }).ToList())
    { }
}

/// <summary>
/// Registry: maps processor type names to factory functions.
/// </summary>
public sealed class Registry
{
    private readonly Dictionary<string, ProcessorFactory> _factories = new();
    private readonly Dictionary<string, ProcessorInfo> _info = new();

    public void Register(ProcessorInfo info, ProcessorFactory factory)
    {
        _factories[info.Name] = factory;
        _info[info.Name] = info;
    }

    public IProcessor Create(string name, ScopedContext ctx, Dictionary<string, string> config)
    {
        return _factories[name](ctx, config);
    }

    public bool Has(string name) => _factories.ContainsKey(name);

    public ProcessorInfo? GetInfo(string name)
        => _info.TryGetValue(name, out var info) ? info : null;

    public List<ProcessorInfo> List() => new(_info.Values);
}

File: zinc-flow-go/zinc-out/fabric/registry/registry.go
package registry

import (
	"zinc-flow/core"
	"fmt"
)

type ProcessorFactory = func(*core.ScopedContext, map[string]string) core.ProcessorFn

//line /home/vrjoshi/proj/caravan-flow/zinc-flow-go/src/fabric/registry/registry.zn:12
type ProcessorInfo struct {
	Name string
	Description string
	ConfigKeys []string
}

func NewProcessorInfo(name string, description string, configKeys []string) ProcessorInfo {
	return ProcessorInfo{Name: name, Description: description, ConfigKeys: configKeys}
}

func (s ProcessorInfo) String() string {
	return fmt.Sprintf("ProcessorInfo(name=%v, description=%v, configKeys=%v)", s.Name, s.Description, s.ConfigKeys)
}

//line /home/vrjoshi/proj/caravan-flow/zinc-flow-go/src/fabric/registry/registry.zn:19
type Registry struct {
	factories map[string]ProcessorFactory
	info map[string]ProcessorInfo
}

func NewRegistry() *Registry {
	return &Registry{factories: map[string]ProcessorFactory{}, info: map[string]ProcessorInfo{}}
}

func (s *Registry) Register(pinfo ProcessorInfo, factory ProcessorFactory) {
	s.factories[pinfo.Name] = factory
	s.info[pinfo.Name] = pinfo
}

func (s *Registry) Create(name string, ctx *core.ScopedContext, config map[string]string) core.ProcessorFn {
	factory := s.factories[name]
	return factory(ctx, config)
}

func (s *Registry) Has(name string) bool {
	return func() bool { _, _ok := s.factories[name]; return _ok }()
}

func (s *Registry) List() []ProcessorInfo {
	result := []ProcessorInfo{}
	names := func() []string { _keys := make([]string, 0, len(s.info)); for _k := range s.info { _keys = append(_keys, _k) }; return _keys }()
	for _, n := range names {
		result = append(result, s.info[n])
	}
	return result
}


